xds/src/main/java/io/grpc/xds/CdsLoadBalancer.java

```diff
@@ -44,77 +41,65 @@
 import io.grpc.xds.internal.sds.TlsContextManagerImpl;
 import java.util.ArrayList;
+import java.util.Collections;
 import java.util.List;
-import java.util.Objects;
-import java.util.concurrent.atomic.AtomicReference;
 import javax.annotation.Nullable;
 
 /**
- * Load balancer for cds_experimental LB policy.
+ * Load balancer for cds_experimental LB policy. One instance per cluster.
  */
-public final class CdsLoadBalancer extends LoadBalancer {
+final class CdsLoadBalancer extends LoadBalancer {
+  // TODO(sanjaypujare): remove once xds security is released
+  @VisibleForTesting
+  static boolean enableSecurity =
+      Boolean.parseBoolean(System.getenv("GRPC_XDS_EXPERIMENTAL_SECURITY_SUPPORT"));
   private final XdsLogger logger;
+  private final LoadBalancer.Helper helper;
   private final LoadBalancerRegistry lbRegistry;
-  private final GracefulSwitchLoadBalancer switchingLoadBalancer;
   private final TlsContextManager tlsContextManager;
-  // TODO(sanjaypujare): remove once xds security is released
-  private boolean enableXdsSecurity;
-  private static final String XDS_SECURITY_ENV_VAR = "GRPC_XDS_EXPERIMENTAL_SECURITY_SUPPORT";
-
-  // The following fields become non-null once handleResolvedAddresses() successfully.
-
-  // Most recent cluster name.
-  @Nullable
   private String clusterName;
-  @Nullable
   private ObjectPool<XdsClient> xdsClientPool;
-  @Nullable
   private XdsClient xdsClient;
+  private ChildLbState childLbState;
 
-  CdsLoadBalancer(Helper helper) {
+  CdsLoadBalancer(LoadBalancer.Helper helper) {
     this(helper, LoadBalancerRegistry.getDefaultRegistry(), TlsContextManagerImpl.getInstance());
   }
 
   @VisibleForTesting
-  CdsLoadBalancer(Helper helper, LoadBalancerRegistry lbRegistry,
+  CdsLoadBalancer(LoadBalancer.Helper helper, LoadBalancerRegistry lbRegistry,
       TlsContextManager tlsContextManager) {
-    checkNotNull(helper, "helper");
+    this.helper = checkNotNull(helper, "helper");
     this.lbRegistry = lbRegistry;
-    this.switchingLoadBalancer = new GracefulSwitchLoadBalancer(helper);
     this.tlsContextManager = tlsContextManager;
     logger = XdsLogger.withLogId(InternalLogId.allocate("cds-lb", helper.getAuthority()));
     logger.log(XdsLogLevel.INFO, "Created");
   }
 
   @Override
   public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
-    logger.log(XdsLogLevel.DEBUG, "Received resolution result: {0}", resolvedAddresses);
-    if (xdsClientPool == null) {
-      xdsClientPool = resolvedAddresses.getAttributes().get(XdsAttributes.XDS_CLIENT_POOL);
-      checkNotNull(xdsClientPool, "missing xDS client pool");
-      xdsClient = xdsClientPool.getObject();
+    if (clusterName != null) {
+      return;
     }
-
+    logger.log(XdsLogLevel.DEBUG, "Received resolution result: {0}", resolvedAddresses);
+    xdsClientPool = resolvedAddresses.getAttributes().get(XdsAttributes.XDS_CLIENT_POOL);
+    xdsClient = xdsClientPool.getObject();
```

Comment:
Hmm, I feel this very redundant and unnecessary. Behavior-wise, they are the same. Checking for null is more helpful for cases you received an argument and save it for later use (e.g., constructor arguments saved as fields and to be used later). Here we are accessing it in-place. If it breaks, it just breaks at the first place where we do not expect it to be null. Regarding the error message, I don't think that provides anything more helpful than the NPE stacktrace containing the line number that breaks it. It is good enough for debugging, which should be very very rare.

Reply:
You can't rely on the line number that breaks if you don't know which commit/version of grpc-java the user is running with.

Reply:
> You can't rely on the line number that breaks if you don't know which commit/version of grpc-java the user is running with.

That is not the usage of `checkNotNull`. Would you put `checkNotNull` for every single variable that is not final? You only want to avoid losing track of the first place the unexpected null value occurs by using `checkNotNull`. Otherwise, its usages are redundant and only makes the program verbose without any benefit. Let it break immediately if it is a programmer error, that's good enough. Doing anything more is excessive.

Why would you want to debug some error without knowing the source code version (aka, what the source code is)?

Reply:
I'm not saying put `checkNotNull` for every single variable that is not final. The `xdsClientPool` is an input from an external component, it's out of the current component's control.

Some times it's hard to know the grpc version from the user reporting issues. Even internally, sometimes you don't know the code was running at which CL or staging or production version.

Reply:
> The xdsClientPool is an input from an external component, it's out of the current component's control.

That's not the necessary reason to check for null. Check if it is null and panic is equivalent to let to throw NPE there.

> Some times it's hard to know the grpc version from the user reporting issues. Even internally, sometimes you don't know the code was running at which CL or staging or production version.

As I said, let it panic at the first place that is unexpected to be null is good enough. Excessive code only adds verbosity.

Reply:
>That's not the necessary reason to check for null. Check if it is null and panic is equivalent to let to throw NPE there.

I'm not saying the behavior is different. My point is helping debug when NPE does happen.

> Excessive code only adds verbosity.

The verbosity helps debug.

Reply:
Let it panic at the line NPE is thrown is good enough for debugging. You do not need the excessive message to find where the exception is thrown. Trying to debug using a different version of source code other than the one the breakage is observed makes no sense.

Reply:
You little clue if you see NPE with a line number of unknown version. If you can see the string "xdsClientPool" in the message of NPE, you immediately know what's going on even you don't know the version.

Reply:
This isn't a big gain while an excessive piece of code is added. If you do this for such reasons, excessive code spreads everywhere.

Reply:
I agree it isn't a big gain. My point is verbosity does help debug, and we have consistently been using that style throughout the code for that purpose. This is not important though so I'm not blocking you on this.

Reply:
I think the checkNotNull should remain where it was and not be changed. There was nothing wrong with it, and it isn't out-of-the-question that it could be null. You can't go around changing existing code based on your whim, because then you get exactly what is happening here: each author has a different approach and the code flip/flops and causes disagreements as to how it should be done.

I'm very bothered about these changes being made in such a sweeping way. Why was all this unnecessary code changed? That just added disagreement opportunities during review and can cause an otherwise-okay change to be hung up on useless stuff.

Reply:
I was cleaning up things alone the way, as we previously had multiple authors added/changed implementations in this class (each of them, as well as reviewers, has very different style and taste), which made the code quite messy. So in this refactoring work, I was trying to polish such things if it can look better (IMO).

Reply:
> So in this refactoring work, I was trying to polish such things if it can look better (IMO).

That's definitely a mistake in two regards: combining two unrelated sizable changes and "IMO".

You're wanting the reviewer to approve changing their code to your code, for no reason other than you think it looks better. And you're "holding them hostage" because they have to choose between the refactoring and having their style remain in-place. It seems it should be obvious why this would go poorly. You should start by assuming the reviewer will not adopt your coding style. Maybe bits and pieces you are able to influence, but y'all will remain different. And then you get to figure out how you work together to form one team.

This PR is sizable at +504 −671. You should have tried to reduce its size in both actual lines changed and non-trivial lines. A small sampling of the non-test code makes it seem like there may be more style changes than refactoring changes. I don't understand why the tests are so dramatically different; I'd have to dig into that more. But just on a cursory glance without actually knowing enough about this code to make a real decision, if I were the reviewer I'd probably "send it back" and have you remove all the superfluous stuff.

Reply:
> I don't understand why the tests are so dramatically different

Rewriting the tests is one of the major goals of this change. For historical reasons (design and implementations have been changing back and forth while each change tried to insert test code even if that made the overall tests worse, and reviewers did not care much about tests), those tests are poorly written if seeing from current's perspective. Really, if you read and try to understand the existing implementation and tests, I bet you would really hate it (e.g., [a quick glance](https://github.com/grpc/grpc-java/blob/5cfbe1061acefc4a859c9fc85fab8c46f9b7a4b1/xds/src/test/java/io/grpc/xds/CdsLoadBalancerTest.java#L340)). It's so messy. But you aren't able to really feel it unless you really dig into it. 😢

Reply:
> Rewriting the tests is one of the major goals of this change.

Okay. I did notice they looked more regular. I don't really understand what "eliminate cluster name change logic" has to do with the tests, but that's fine. As long as you two know what's going on and are happy SGTM.

Reply:
But I believe unrelated changes should be made in separate PRs.

Reply:
Ah, I can separate out the test class into a separate PR, that's a complete rewritten thing while github diff is just trying hard to retain the very little common code. I believe tests in this PR should also work for things before refactoring (maybe except the portion of security code) as it's blackbox. But reviewers will basically still see similar diffs, but most just contain the test class. 

Regardless the tests. Really this PR only refactors the implementation structure to get rid of some unnecessary logic. But since it's a structural refactoring (aka, reimplement a couple of inner components), the change is inevitably big. But other than that, remaining is just cosmetic things (e.g., Sanjay's `enableXdsSecurity` flag), if you really think those are "unrelated changes".

You wouldn't feel how painful it was to deal with the existing code, especially the test code. That's why rewriting is way better than trying to refactor while keep change minimal.

Reply:
I'm not bothered by the massive changes to the tests. That's not uncommon in tests. It was clear to me they were rewritten and the diff was unhelpful.

I'm bothered by refactoring, API-unrelated test changes, and "cosmetic" changes done together. It sounds like the refactoring by itself would have been a small change to make and review (but yes, it'd have many changed lines). And testing changes may have been gladly received and reviewed. But it is clear that this PR description says it is for the refactoring. If the tests are so important, why weren't they called out at all?

I'm actually not _too_ bothered with sending this all out as a single PR. I think it was a clear mistake, but hey, it happens and I know others don't quite feel the same way I do about how to refactor. I think a better thing to realize is that when the review devolved, the approach should have changed to removing the contentious-changes-that-weren't-the-core-of-the-change and getting the "eliminate cluster name change logic in CDS LB policy" part merged. Digging your heals in and arguing the changes were good is unhelpful to the objective of the PR (as defined by its title and description). I can understand that these may have been long-standing disagreements, but _we can make progress_ even with the disagreements.